xarray/core/dataset.py

```diff
@@ -3150,11 +3150,25 @@ def drop(self, labels, dim=None, *, errors='raise'):
             any of the variable or index labels passed are not
             in the dataset. If 'ignore', any given labels that are in the
             dataset are dropped and no error is raised.
+        **dim_kwargs : {dim: label, ...}, optional
```

Comment:
Everywhere else in xarray where we support keyword arguments, we try to use the `xarray.core.utils. either_dict_or_kwargs()` helper function, to allow writing either `dataset.method({k: v})` or `dataset.method(k=v)` inter-changeably.

It would be nice to do the same here, if possible, though it get a little bit tricky due to the need to support the original signature. I think something like this would work:
```python
if dim_kwargs or utils.is_dict_like(labels):
    dim_kwargs = utils.either_dict_or_kwargs(labels, dim_kwargs, 'drop')
    if dim is not None:
        raise ValueError(...)
    ...
```

Reply:
Isn't `labels` supposed to be `list`-like, not `dict`-like? That is why I did not try to use `either_dict_or_kwargs`.

Reply:
I'm proposing allowing labels to be *either* `list`-like or `dict`-like.

Though perhaps that's more confusing...

Reply:
Ah, I see. This means we would also want to support the user dropping labels along multiple dimensions? The docstring would be complicated. You also mentioned above that "we might want to deprecate the current function signature", but the `dict`-like approach does not allow `dim=None`. This means we would need to support four different ways of calling the function:

- `labels` is `list`-like but no `dim`: drop variables.
- `labels` is `list`-like and `dim`: drop labels along `dim`.
- `labels` is a `dict`: drop values for each `dim` key along that dimension.
- `labels_kwargs` are used: drop values for each `dim` keyword argument.

As a first-time contributor, I don't think this is my call, but I'd vote for deprecating the "drop variables" approach in favor of `dict`-like long-term to be more consistent with other functions.

Maybe we add a new function for dropping variables or an argument `variables`?

Reply:
In the long term, can we deprecate 
> labels is list-like and dim: drop labels along dim.

and then have 3 modes of operation? i.e. when `labels` is list-like, they must be variable names.

Reply:
Do you think that's more clear for a user than `drop(..., variables=True)`? IMO explicit is better than implicit, and conditioning on the type of the input seems confusing, especially when that input type was previously valid. In the worst case, this would fail silently by dropping variables when a user meant dimensions.

Reply:
Sorry, wrote a confused reply. Yeah, I like your idea. Shall I implement?

Reply:
> In the long term, can we deprecate
> 
> > labels is list-like and dim: drop labels along dim.
> 
> and then have 3 modes of operation? i.e. when `labels` is list-like, they must be variable names.

I think this would be make sense. We'll need to support this through a deprecation cycle, but eventually we can remove it.

Reply:
Great. I followed another bit of code that just used `warnings.warn` with a `DeprecationWarning`.